File: cpp/src/solutions/string_to_integer_atoi.cpp
// LeetCode 8. String to Integer (atoi)

#include "cpp/src/common/common.h"

class Solution {
 public:
  int myAtoi(string str) {
    int i = 0;
    while (i < str.size() && str[i] == ' ') {
      i++;
    }
    bool positive = true;
    if (i < str.size() && (str[i] == '+' || str[i] == '-')) {
      positive = str[i] == '+';
      i++;
    }
    long long ans = 0;
    while (i < str.size()) {
      auto ch = str[i];
      if (!isdigit(ch)) {
        break;
      }
      auto bit = positive ? ch - '0' : '0' - ch;
      ans = ans * 10 + bit;
      if (ans > numeric_limits<int>::max()) {
        ans = numeric_limits<int>::max();
      }
      if (ans < numeric_limits<int>::min()) {
        ans = numeric_limits<int>::min();
      }
      i++;
    }
    return ans;
  }
};

TEST(myAtoi, example1) {
  string str = "42";
  int ans = 42;
  EXPECT_EQ(Solution().myAtoi(str), ans);
}

TEST(myAtoi, example2) {
  string str = "-42";
  int ans = -42;
  EXPECT_EQ(Solution().myAtoi(str), ans);
}

TEST(myAtoi, example3) {
  string str = "4193 with words";
  int ans = 4193;
  EXPECT_EQ(Solution().myAtoi(str), ans);
}

TEST(myAtoi, example4) {
  string str = "words and 987";
  int ans = 0;
  EXPECT_EQ(Solution().myAtoi(str), ans);
}

TEST(myAtoi, example5) {
  string str = "-91283472332";
  int ans = -2147483648;
  EXPECT_EQ(Solution().myAtoi(str), ans);
}

File: rust/src/solutions/string_to_integer_atoi.rs
// LeetCode 8. String to Integer (atoi)

use super::Solution;

impl Solution {
    pub fn my_atoi(s: String) -> i32 {
        let mut chars = s.trim_start().bytes().peekable();
        let positive = match chars.peek() {
            Some(b'+') => {
                chars.next();
                true
            }
            Some(b'-') => {
                chars.next();
                false
            }
            _ => true,
        };
        let mut ans: i32 = 0;
        for ch in chars {
            if ch.is_ascii_digit() {
                let bit = ch - b'0';
                ans = ans.saturating_mul(10);
                ans = if positive {
                    ans.saturating_add(bit.into())
                } else {
                    ans.saturating_sub(bit.into())
                };
            } else {
                break;
            }
        }
        ans
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn example1() {
        let s = String::from("42");
        let ans = 42;
        assert_eq!(Solution::my_atoi(s), ans);
    }

    #[test]
    fn example2() {
        let s = String::from("-42");
        let ans = -42;
        assert_eq!(Solution::my_atoi(s), ans);
    }

    #[test]
    fn example3() {
        let s = String::from("4193 with words");
        let ans = 4193;
        assert_eq!(Solution::my_atoi(s), ans);
    }

    #[test]
    fn example4() {
        let s = String::from("words and 987");
        let ans = 0;
        assert_eq!(Solution::my_atoi(s), ans);
    }

    #[test]
    fn example5() {
        let s = String::from("-91283472332");
        let ans = -2147483648;
        assert_eq!(Solution::my_atoi(s), ans);
    }
}
